IOS-9356 Adjust crouton constraints above the tabbar or the safe area

L-Trujillo26 opened the pull request:
<!-- This warning is to remind you to check if a designer needs to review this PR. Delete it when you consider it -->
> [!WARNING]
> Should a designer need to review/verify this PR?

## 🎟️ **Jira ticket**
[IOS-9356](https://jira.tid.es/browse/IOS-9356)
<img width="887" alt="image-2024-11-07-16-49-23-770" src="https://github.com/user-attachments/assets/b575db50-cc08-4aeb-8cea-8a313ce47e33" />


## 🥅 **What's the goal?**

The  Crouton design is deprecated and replaced with a new design that matches the appearance and behavior of the snackbar, as specified in the Figma:

https://www.figma.com/design/yCHLIfy4WMfRdlADwyL4kZ/%F0%9F%94%B8-Snackbar-Specs?t=bb57Dv8A0XsJVGqF-0


The component name has not yet been changed, but a floating design is implemented that respects the safe area and elements like the TabBar ensuring proper display in the view hierarchy.

## 🚧 **How do we do it?**
Design changes have been made(behavior is aligns with the Figma specifications)

## 🧪 **How can I verify this?**
MisticaCatalog -> Crouton (UIKit)-> show Crouton


This screenshot shows the Snackbar in SwiftUI and the Crouton in UIKit.
<img width="708" alt="crouton" src="https://github.com/user-attachments/assets/37c3bc9f-767b-4fec-9c6e-2913fe0fb83e" />


https://github.com/user-attachments/assets/e33487ae-0d8f-44c9-985c-f6cfb6e39785

## 🏑 **AppCenter build**

https://appcenter.ms/orgs/Tuenti-Organization/apps/Mistica-SwiftUI-iOS/distribute/releases

L-Trujillo26 commented:
/record-screenshots

amegias commented:
An extra point:
I would remove "all brands" from existing Crouton tests. IMHO, it is more difficult to verify screenshots in the PR. wdyt David/Laura?

L-Trujillo26 commented on Sources/Mistica/Components/Crouton/Presentation/CroutonView.swift:
```diff
@@ -242,53 +243,55 @@ private extension CroutonView {
 
-            directionalLayoutMargins = Constants.margins
+    private func findTabBar(in view: UIView) -> UITabBar? {
+        for subview in view.subviews {
```
I had a logic to check if a tabBar exists and if it is located at the bottom of the device. If this condition isn't met, it would continue searching. I am going to add again this logic to prevent cases like this.

L-Trujillo26 replied:
done and added tests

DavidMarinCalleja commented:
> An extra point: I would remove "all brands" from existing Crouton tests. IMHO, it is more difficult to verify screenshots in the PR. wdyt David/Laura?

I don't really have a strong opinion about this because testing “all brand” increases the size of the repo, makes it difficult to check the PR but I'm afraid that removing it will hide some bug. But if you agree we can remove it.

amegias commented:
> > An extra point: I would remove "all brands" from existing Crouton tests. IMHO, it is more difficult to verify screenshots in the PR. wdyt David/Laura?
> 
> I don't really have a strong opinion about this because testing “all brand” increases the size of the repo, makes it difficult to check the PR but I'm afraid that removing it will hide some bug. But if you agree we can remove it.

then, we should add test to the palette, etc

DavidMarinCalleja commented on Sources/Mistica/Components/Crouton/Presentation/CroutonView.swift:
@L-Trujillo26 I have a doubt, where the crouton is fixed at the end? I mean the view is added to the current view and therefore disappears if we navigate to another site or on the other hand is fixed to the root view of the app and therefore the crouton is fixed even after browsing.

L-Trujillo26 replied:
If a exact UIViewController isn't specified, the crouton will be add to the RootViewController making it visible during navigation.  Its persistence depends on the configuration set in the initialization. This behavior has been confirmed with the design team.

DavidMarinCalleja reviewed (approved):
<img src="https://media2.giphy.com/media/aLdiZJmmx4OVW/giphy.gif?cid=bd3ea57ezafxnhd5l19cv4hcniiwaq2rgrhd1nt9iie9ljte&ep=v1_gifs_search&rid=giphy.gif&ct=g"/>
Great job!!

amegias commented:
I've created a pet project to test this from UIKit.
It works with:
- TabBar
- TabBar where tab1 presents a modal
- TabBar where tab1 push a VC hidding the tabBar
- TabBar where tab1 push a VC keeping the tabBar
- ViewController

I just found this glitch, could you check it please:

https://github.com/user-attachments/assets/7e4a3d47-fafa-4a9c-aa05-c129b429f397

L-Trujillo26 commented:
> I've created a pet project to test this from UIKit. It works with:
> 
> * TabBar
> * TabBar where tab1 presents a modal
> * TabBar where tab1 push a VC hidding the tabBar
> * TabBar where tab1 push a VC keeping the tabBar
> * ViewController
> 
> I just found this glitch, could you check it please:
> 
>  bug_dismiss.mp4

This is the animation for the crouton positioned at the bottom. I have updated it to the snackbar animation (it fades in and out gradually).

Attached is a video of the Snackbar animation in SwiftUI and UIKit.

https://github.com/user-attachments/assets/71ff613d-3c15-4f00-b45c-7f4e00e66e47

L-Trujillo26 commented:
> I've created a pet project to test this from UIKit. It works with:
> 
> * TabBar
> * TabBar where tab1 presents a modal
> * TabBar where tab1 push a VC hidding the tabBar
> * TabBar where tab1 push a VC keeping the tabBar
> * ViewController
> 
> I just found this glitch, could you check it please:
> 
>  bug_dismiss.mp4

This is the animation for the crouton positioned at the bottom. I have updated it to match the snackbar animation (it fades in and out gradually).

Attached video of the Snackbar animation in SwiftUI and UIKit.


https://github.com/user-attachments/assets/1c6fd404-234d-4715-bd6b-c0948285a1a4

amegias commented on Sources/Mistica/Components/Crouton/Presentation/CroutonView.swift:
```diff
@@ -212,12 +217,15 @@ extension CroutonView {
-            delay: Constants.contentAnimationDelay,
-            options: .curveEaseIn,
+            delay: 0,
+            options: .curveEaseInOut,
```
remove them since they are the default values

amegias commented on Sources/Mistica/Components/Crouton/Presentation/CroutonView.swift:
```diff
@@ -177,14 +176,20 @@ extension CroutonView {
-            delay: Constants.contentAnimationDelay,
-            options: .curveEaseOut,
+            delay: 0,
+            options: .curveEaseInOut,
```
remove them since they are the default values

amegias commented on Sources/Mistica/Components/Crouton/Presentation/CroutonView.swift:
```diff
@@ -212,12 +217,15 @@ extension CroutonView {
             animations: {
                 self.alpha = 0
+                self.bottomConstraint?.constant = self.frameHeight + abs(originalBottomConstant)
+                superview.layoutIfNeeded()
```
superview? why?

L-Trujillo26 commented:
> > I've created a pet project to test this from UIKit. It works with:
> > 
> > * TabBar
> > * TabBar where tab1 presents a modal
> > * TabBar where tab1 push a VC hidding the tabBar
> > * TabBar where tab1 push a VC keeping the tabBar
> > * ViewController
> > 
> > I just found this glitch, could you check it please:
> > bug_dismiss.mp4
> 
> This is the animation for the crouton positioned at the bottom. I have updated it to match the snackbar animation (it fades in and out gradually).
> 
> Attached video of the Snackbar animation in SwiftUI and UIKit.
> 
>  animationSnackbar.mov

Changed animation to spec(https://www.figma.com/design/yCHLIfy4WMfRdlADwyL4kZ/%F0%9F%94%B8-Snackbar-Specs?node-id=2830-373&t=G3hOMQak63nhBH7R-4)

https://github.com/user-attachments/assets/c1d98996-30a4-4803-9db2-650b3cc37a01

amegias commented on Sources/Mistica/Components/Crouton/Presentation/CroutonView.swift:
```diff
@@ -177,14 +176,20 @@ extension CroutonView {
             animations: {
                 self.alpha = 1
+                self.bottomConstraint?.constant = originalBottomConstant ?? 0
+                container.layoutIfNeeded()
```
why?

yceballost reviewed (approved):
perfect! ✨
